### PR TITLE
T016 - Protein-ligand interactions

### PR DESCRIPTION
**Pull request by t-kimber**
* Talktorial ID: T016
* Title: protein-ligand interactions
* Original authors: Michele Wichmann, Jaime Rodriguez-Guerra, Dominique Sydow
* Reviewer(s): Talia B. Kimber, Yonghui Chen, Andrea Volkamer
* Date of review: 14.12.2020

# Content review

* Potential labels or categories (e.g. machine learning, small molecules, online APIs): XXXXXXXX
* One line summary: XXXXXXXXXX
* [x] The table of contents reflects the talktorial story-line; order of #, ##, ### headers is correct
* [x] URLs are linked with meaningful words, instead of pasting the URL directly or linking words like `here`.
* [x] I have spell-checked the notebook
* [x] Images have enough resolution to be rendered with quality, without being _too_ heavy.
* [x] All figures have a description
* [x] Markdown cell content is still in-line with code cell output (whenever results are discussed)
* [x] I have checked that cell outputs are not incredibly long (this applies also to `DataFrames`)
* [x] Formatting looks correctly on the Sphinx render (bold, italics, figure placing)

# Code review

* Time it took to execute (approx.):
* [x] Variable and function names follow snake case rules (e.g. `a_variable_name` vs `aVariableName`)
* [x] Spacing follows PEP8 (run Black on the code cells if needed)
* [x] Code line are under 99 characters each (run `black -l 99`)
* [x] Comments are useful and well placed
* [x] There are no unpythonic idioms like `for i in range(len(list))` (see slides)
* [x] All 3rd party dependencies are listed at the top of the notebook
* [x] I have marked all code cell with output referenced in markdown cells with the label `# TODO: CI`
* [x] I have identified potential candidates for a code refactor / useful functions
* [x] All `import ...` lines are at the top (practice part) cell, ordered by standard library / 3rd party packages / our own (`teachopencadd.*`)
* [x] I have update the relative paths to absolute paths.
* List here unfamiliar libraries you find in the imports and their intended use:

For abs vs rel paths, use these:

```python
HERE = Path(_dh[-1])
DATA = HERE / "data"
```

### CODE REVIEW

File: teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb [hunk 1]
```diff
@@ -4,58 +4,83 @@
    "cell_type": "markdown",
```
Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>Maybe start with sth more in the direction of the below quote: understanding protein–ligand interactions which are driving molecular recognition are fundamental in (rational) drug design.</p><p>Then that we use PLIP (and NGLView) to do so (maybe link PLIP already here?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>I feel like we don't need this quote, can you extract the important numbers (/study design) and mention them shortly in the figure caption instead? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <blockquote><a href="https://plip.biotec.tu-dresden.de/plip-web/plip/index" target="_blank">PLIP</a>&nbsp;is one of the most popular thanks to its publicly available webserver and free-to-use Python library</blockquote><p>-&gt; a popular (not sure if 'most' is necessarily correct)</p><p> </p><p>I would prefer if instead of quoting the paper here, we would have a bit more (maybe bullet point like) explanation in our own words, listing the interactions and the general idea of how they are calculated</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>Maybe we can find a more overview like figure that combines the different interactions? The individual figures are linked in the pdf already anyways (btw I really like the showing of the pdf in the nb!!!)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>Just wondering if we could include this in the environment file? (tbd)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>Maybe this could be a title and used in the pratical description above?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>side note for NGLview, Jaime shared recently with me, to center the figures on the binding site per default one can use this: <code>v.center("ligand")</code> (see <a href="http://nglviewer.org/ngl/api/manual/selection-language.html#expressions" target="_blank">manual</a>)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>This could be moved above to where you installed it, no?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <blockquote>PLIP only accepts PDB files, so we can pass the PDB file </blockquote><p>'only' here sounds like this is unfavorable for our example, maybe rephrase</p><p> </p><blockquote> processes each detected binding site</blockquote><p>How are binding sites 'detected'? I think PLIP uses all ligands (HETATM ids) in the PDB File and all residues within xA from them (or so)? Could you double check and add this info?</p><p> </p><p>[coming back to a comment in the theory, maybe the bullet points could be the same there, with short descriptions behind the colon]</p><p> </p><p>Not sure it is full clear here what the report contains: one entry per interaction, or per residue? Maybe an example helps?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>This function is supercool, but also quite complex, I'm wondering if it is understandable for beginners.</p><ul><li>Maybe if the reports are introduced in more detail above it becomes clear.</li><li>Also what could help might be a bit textual (markdown explanation) above the function?</li></ul><p>tbd</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>I assume this is an open 'TODO', for now I didn't run the nb, will get back to this and see if I can help ones i ran it locally.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>This description of the dictionary is very clear :) Maybe move this info up?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>Could we show the 'main' ligand here? The users are most likely less interested in the sulfate ion, than in the ATP competitive ligand... We could comment on that in general here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>shouldn't the color_map be a parameter of the function?</p><p> </p><p>Maybe also her you could add a few comments and/or a bit more information on the function above? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>[Again I didn't run it locally yet, so I don't see the actual figure]</p><p> </p><p>Anyhow could you add a few sentences on what we see (probably since it is a kinase, the ligand shows 1 (or 2 or 3) Hbonds to the hinge region, and maybe some hydrophobic backpocket interactions or so.</p><p> </p><p>And one could link to KLIFS <a href="https://klifs.net/details.php?structure_id=7308" target="_blank">3poz</a> and compare the main findings to the KLIFS interaction fingerprint.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

**Review by AndreaVolkamer — COMMENTED** [review]
@t-kimber the talktorial looks great! I added a few comments on putting more info - or the info that is there in a slightly different format or position.

File: teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb (hunk 1 above)
Comment (AndreaVolkamer) [reply teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>Oh actually how about we simply use the figure for 3poz (or a protein of your choice) when using the website? Maybe also the tables they show can be reused in the talktorial for explanation?</p>

File: teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb [hunk 2]
```diff
@@ -4,201 +4,186 @@
...
```
Comment (yonghui-cc) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>I suggest making this (figure 1) smaller a bit.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>This module can't import because of the current python path.</p><p>There is an idea on my computer:</p><pre class="ql-syntax" spellcheck="false">import sys
sys.path.append('../../..')
from teachopencadd.utils import show_pdf
</pre><p>Maybe you have a better idea.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>I suggest use:</p><pre class="ql-syntax" spellcheck="false">create_df_from_binding_site(list(interactions_by_site.keys())[0],
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;interaction_type="hydrophobic")
</pre><p>to make the code more readable.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <ol><li>Could you show the list of the colors corresponding to the interaction type, at least write something with Markdown text to explain the color?  It even better if the name of the residues could be shown.</li><li>I also saw some interactions which are not between protein and ligand, should we remove them?</li></ol><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <ol><li>Also halogen interactions with THR790, MET766</li><li>a small question: Does there need space between the residues(LYS) and the number(745)?</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: <p>I suggest splitting this cell into two cells</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (t-kimber) [reply teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb:4]: Resolved by updating environment.

**Comment by AndreaVolkamer** [comment]
@t-kimber well done. I only changed small things. Let me know if you agree. 
@jaimergp one open question for you: In the website the nglview image that is rendered in the notebook does not show (in my local set-up it also only shows after executing the cell twice). Is there an easy fix for this?
Besides that T016 would be ready for being merged from my side.

**Comment by t-kimber** [comment]
Thank you for your final review @AndreaVolkamer !
Concerning nglview rendering, there seems to be issues with the latest version of jupyter lab (v.3) (I forced jupyter lab 2 locally). Since most notebooks in the revamp are concerned, the changes will be done with a PR targeting main. There is an open issue https://github.com/volkamerlab/teachopencadd/issues/80

**Comment by AndreaVolkamer** [comment]
Great, ready to be merged then. @jaimergp can you do the merge please.
